lib/view_model/access_control/composed.rb

```diff
@@ -0,0 +1,200 @@
+
+    def error
+      case
+      when veto;   veto_error
+      when !allow; allow_error
+      else;        nil
+      end
+    end
+
+    # Merge this composed result with another. `allow`s widen and `veto`es narrow.
+    def merge(&block)
+      if self.veto
+        self
+      else
+        other = yield
+
+        new_allow = self.allow || other.allow
+
+        new_allow_error =
+          case
+          when new_allow
+            nil
+          when !self.allow_error
+            other.allow_error
+          when !other.allow_error
+            self.allow_error
+          else
+            self.allow_error.merge(other.allow_error)
+          end
+
+        ComposedResult.new(new_allow, other.veto, new_allow_error, other.veto_error)
+      end
+    end
+  end
+
+  # Error type when no `if` conditions succeed.
+  class NoRequiredConditionsError < ViewModel::AccessControlError
+    attr_reader :reasons
+
+    def initialize(nodes, reasons)
+      super("Action not permitted because none of the possible conditions were met.", nodes)
+      @reasons = reasons
+    end
+
+    def metadata
+      super.merge(conditions: @reasons.to_a)
+    end
+
+    def merge(other)
+      NoRequiredConditionsError.new(nodes | other.nodes,
+                                    Lazily.concat(reasons, other.reasons).uniq)
+    end
+  end
+
+  def initialize(location, ancestors: nil)
+    super()
+
+    @ancestors = Array.wrap(ancestors)
+    @location  = location
+
+    @edit_valid_ifs      = []
+    @edit_valid_unlesses = []
+
+    @editable_ifs        = []
+    @editable_unlesses   = []
+
+    @visible_ifs         = []
+    @visible_unlesses    = []
+  end
+
+  def visible_if!(reason, &block)
+    @visible_ifs         << new_permission_check(reason, &block)
+  end
+
+  def visible_unless!(reason, &block)
+    @visible_unlesses    << new_permission_check(reason, &block)
+  end
+
+  def editable_if!(reason, &block)
+    @editable_ifs        << new_permission_check(reason, &block)
+  end
+
+  def editable_unless!(reason, &block)
+    @editable_unlesses   << new_permission_check(reason, &block)
+  end
+
+  def edit_valid_if!(reason, &block)
+    @edit_valid_ifs      << new_permission_check(reason, &block)
+  end
+
+  def edit_valid_unless!(reason, &block)
+    @edit_valid_unlesses << new_permission_check(reason, &block)
+  end
+
+  # final
+  def visible_check(view, context:)
+    env = ViewEnv.new(view, context)
+    check_delegates(env, visible_ifs, visible_unlesses)
+  end
+
+  # final
+  def editable_check(view, deserialize_context:)
+    env = EditEnv.new(view, deserialize_context)
+    check_delegates(env, editable_ifs, editable_unlesses)
+  end
+
+  # final
+  def valid_edit_check(view, deserialize_context:, changes:)
+    env = EditEnv.new(view, deserialize_context, changes)
+    check_delegates(env, edit_valid_ifs, edit_valid_unlesses)
+  end
+
+  protected
+
+  def visible_ifs
+    Lazily.concat(@visible_ifs,         *ancestors.map(&:visible_ifs))
```

Comment:
👍  I think, but noticed another issue: the parent's `always` ancestor shows up multiple times.
`["child tree1", "child always", "parent always", "parent tree1", "parent always"]` I could keep a set and avoid visiting twice, but now I'm allocating again.